WcaOnRails/app/webpacker/components/CompetitionOverview/CompetitionTable.js

```diff
@@ -0,0 +1,110 @@
+
+  if (mode === 'future') {
+    const msDifference = startDate.getTime() - dateToday.getTime();
+    const dayDifference = Math.ceil(msDifference / (1000 * 3600 * 24));
+    return dayDifference;
+  }
+  if (mode === 'past') {
+    const msDifference = dateToday.getTime() - endDate.getTime();
+    const dayDifference = Math.floor(msDifference / (1000 * 3600 * 24));
+    return dayDifference;
+  }
+
+  return -1;
+}
+
+function renderYearHeader(competitions, index, sortByAnnouncement) {
+  if (index > 0 && competitions[index].year !== competitions[index - 1].year
+    && !sortByAnnouncement) {
+    return <li className="list-group-item break">{competitions[index].year}</li>;
+  }
+  return null;
+}
+
+function renderRegistrationStatus(comp) {
+  if (comp.registration_status === 'not_yet_opened') {
+    return <i className="icon clock blue" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.registration.opens_in', { duration: comp.timeUntilRegistration })} />;
+  }
+  if (comp.registration_status === 'past') {
+    return <i className="icon user times red" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.registration.closed', { days: I18n.t('common.days', { count: calculateDate(comp, 'future') }) })} />;
+  }
+  if (comp.registration_status === 'full') {
+    return <i className="icon user clock orange" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.registration.full')} />;
+  }
+
+  return <i className="icon user plus green" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.registration.open')} />;
+}
+
+function renderDateIcon(comp, showRegistrationStatus, sortByAnnouncement) {
+  if (comp.isProbablyOver) {
+    if (comp.resultsPosted) {
+      return <i className="icon check circle result-posted-indicator" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.hourglass.posted')} />;
+    }
+
+    return <i className="icon hourglass end" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.hourglass.ended', { days: I18n.t('common.days', { count: calculateDate(comp, 'past') }) })} />;
+  }
+  if (comp.inProgress) {
+    return <i className="icon hourglass half" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.hourglass.in_progress')} />;
+  }
+  if (sortByAnnouncement) {
+    return <i className="icon hourglass start" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.hourglass.announced_on', { announcement_date: comp.announcedDate })} />;
+  }
+  if (showRegistrationStatus) {
+    return renderRegistrationStatus(comp);
+  }
+
+  return <i className="icon hourglass end" data-toggle="tooltip" data-original-title={I18n.t('competitions.index.tooltips.hourglass.starts_in', { days: I18n.t('common.days', { count: calculateDate(comp, 'future') }) })} />;
+}
+
+function CompetitionTable({
```

Comment:
I think it's standard to have the main component/function at the top, for quick access to see what this component is about and so that function dependencies point downward.

Reply:
I can't pass the pre-commit hooks with this because ESLint isn't happy with it... `no-use-before-define`. Should I try circumventing this?

Reply:
Yep you're right the problem was the function calls. I turned things into components and this worked.